x/logic/predicate/file.go

```diff
@@ -0,0 +1,93 @@
+package predicate
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sort"
+
+	"github.com/ichiban/prolog/engine"
+)
+
+// SourceFile is a predicate that unify the given term with the currently loaded source file. The signature is as follows:
+//
+// source_file(?File).
+//
+// Where File represents a loaded source file.
+//
+// Example:
+//
+//	# Query all the loaded source files.
+//	- source_file(File).
+//
+//	# Query the given source file is loaded.
+//	- source_file('foo.pl').
+func SourceFile(vm *engine.VM, file engine.Term, cont engine.Cont, env *engine.Env) *engine.Promise {
+	loaded := getLoadedSources(vm)
+
+	inputFile, err := getFile(env, file)
+	if err != nil {
+		return engine.Error(fmt.Errorf("source_file/1: %w", err))
+	}
+
+	if inputFile != nil {
+		if _, ok := loaded[*inputFile]; ok {
+			return engine.Unify(vm, file, engine.NewAtom(*inputFile), cont, env)
+		}
+		return engine.Delay()
+	}
+
+	promises := make([]func(ctx context.Context) *engine.Promise, 0, len(loaded))
+	sortedSource := sortLoadedSources(loaded)
+	for i := range sortedSource {
+		term := engine.NewAtom(sortedSource[i])
+		promises = append(
+			promises,
+			func(ctx context.Context) *engine.Promise {
+				return engine.Unify(
+					vm,
+					file,
+					term,
+					cont,
+					env,
+				)
+			})
+	}
+
+	return engine.Delay(promises...)
+}
+
+func getLoadedSources(vm *engine.VM) map[string]interface{} {
+	loadedField := reflect.ValueOf(vm).Elem().FieldByName("loaded").MapKeys()
+	loaded := make(map[string]interface{}, len(loadedField))
+	for _, value := range loadedField {
+		loaded[value.String()] = nil
```

Comment:
I usually use a `map[string]struct{}` to represent sets, with `struct{}{}` as an `empty` value (unique and named). But this is a matter of taste.